File: Python/Location.py
from itertools import combinations
import random
import math
import copy
from AdjacenceMatrix import *

class Location:
    def __init__(self, lati, longi):
        self.latitude = lati
        self.longitude = longi

    def __repr__(self):
        return self.__str__()

    def __str__(self):
        return f"[{self.latitude} : {self.longitude}]"

    def __eq__(self, obj):
        return isinstance(obj, Location) and obj.latitude == self.latitude and obj.longitude == self.longitude 

    def __hash__(self) -> int:
        return self.latitude + self.longitude * 10000000

def generate(n):
    locs = []
    for _ in range(n):
        longi = random.randint(0, 1000)
        lati = random.randint(0, 1000)
        locs.append(Location(lati, longi))
    return locs


def distance(location1, location2):
    longi = location1.longitude - location2.longitude
    lati = location1.latitude - location2.latitude
    return math.sqrt(lati**2 + longi**2)

def cluster_distance(locations1: list, locations2: list) -> float:
    d = distance(locations1[0], locations2[0])
    for l in locations1:
        for g in locations2:
            d = min(d, distance(l, g))
    return d

def closest_elements(locations1: list, locations2: list) -> tuple:
    t = (locations1[0], locations2[0])
    d = distance(locations1[0], locations2[0])
    for l in locations1:
        for g in locations2:
            if d > distance(l, g):
                d = distance(l, g)
                t = (l, g)
    d = min(distance(l,g) for l in locations1 for g in locations2)
    return t

def diameter(locations: list) -> float:
    if len(locations) == 1:
        return 0
    comb = list(combinations(locations, 2))
    ds = [distance(c[0], c[1]) for c in comb]
    return max(ds)

def clustering_distances(locations: list) -> list:
    loc = copy.copy(locations)
    clus = [loc[0]]
    loc.remove(loc[0])
    distances = []
    while loc:
        t = closest_elements(clus, loc)
        clus.append(t[1])
        loc.remove(t[1])
        distances.append(distance(t[0], t[1]))
    distances.sort()
    return distances

def center(coll: list) -> Location:
    return Location(sum(x.latitude for x in coll)/len(coll), sum(x.longitude for x in coll)/len(coll))


def overlaps(coll) -> bool:
    a = [True for i in coll for j in coll if cluster_distance(i, j) == 0]
    return bool(a)

def merge_overlapping_sets(lsts):
    repeat = True
    while repeat:
        repeat = False
        for c in lsts:
            for e in c:
                r = [l for l in lsts if e in l]
                if len(r) > 1:
                    for m in r:
                        lsts.remove(m)
                    tt = [el for sublist in r for el in sublist]
                    lsts.append(tt)
                    repeat = True
                    break
            if repeat:
                break


def splitting(coll, n) -> list:
        stair = clustering_distances(copy.copy(coll))
        dis = stair[- n + 1]
        lsts = [[coll[0]]]
        for a in coll[1:]:
            appended = False
            for e in lsts:
                if cluster_distance([a], e) < dis:
                    e.append(a)
                    appended = True
            if not appended:
                lsts.append([a])
        merge_overlapping_sets(lsts)
        if len(lsts) != n:
            raise ValueError(f"The splitting produced {len(lsts)} instead of {n}\nThe collections are\n{lsts}")
        return lsts

def clustering(coll, n) -> list:
    strt = random.choices(coll, k= n)
    clusters = []
    for i in range(4):
        clusters = [[e] for e in strt]
        for e in coll:
            min(clusters, key=lambda x: cluster_distance([e], x)).append(e)
        if i > 0:
            for index, c in enumerate(strt):
                clusters[index].remove(c)
                if len(clusters[index]) == 0:
                    raise ValueError(f"removed last element {c} from one collection {clusters}")
        strt = [center(cluster) for cluster in clusters]
    return clusters

File: Julia/Location.jl
using Test
using Random

struct Location
    longitude::Int
    latitude::Int
end

function distance(location::Location, otherlocation::Location)
    return sqrt((location.longitude - otherlocation.longitude)^2 + (location.latitude - otherlocation.latitude)^2)
end

function distance(location::Location, otherlocations::Vector{Location})
    return min(map(x -> distance(location, x), otherlocations)...)
end

function distance(otherlocations::Vector{Location}, location::Location)
    return distance(location, otherlocations)
end

function distance(locations::Vector{Location}, otherlocations::Vector{Location})
    return min(map(x -> distance(locations, x), otherlocations)...)
end

function closestelements(locations, otherlocations)::Vector{Location}
    closestelements = [locations[1], otherlocations[1]]
    dst = distance(closestelements...)
    for loc in locations
        for otherloc in otherlocations
            if dst > distance(loc, otherloc)
                closestelements = [loc, otherloc]
                dst = distance(loc, otherloc)
            end
        end
    end
    return closestelements
end

function pathlength(locations::Vector{Location})::Float64
    tot = 0
    for i = 1:(length(locations)-1)
        tot += distance(locations[i], locations[i+1])
    end
    return tot
end

function clusteringdistances(locations)
    otherlocations = copy(locations)
    cluster = [pop!(otherlocations)]
    distances = []
    while !isempty(otherlocations)
        pair = closestelements(cluster, otherlocations)
        push!(cluster, pair[2])
        deleteat!(otherlocations, findall(x->x==pair[2], otherlocations))
        push!(distances, distance(pair[1], pair[2]))
    end
    sort!(distances)
    return distances        
end

function clustering(locations::Vector{Location}, n)
    stair = clusteringdistances(locations)
    dist = stair[end-n + 1]
    lsts::Vector{Vector{Location}} = [[locations[1]]]
    for location in locations
        if location == locations[1]
            continue
        end
        idxs = findall(x->distance(location, x) <= dist, lsts)
        mergedcollections = mergecollections(lsts, idxs)
        push!(mergedcollections, location)
        deleteat!(lsts, idxs)
        push!(lsts, mergedcollections)
    end
    return lsts
end

function mergecollections(collection::Vector{Vector{Location}}, indexes)::Vector{Location}
    mergedcollections = []
    for i in indexes
        append!(mergedcollections, collection[i])
    end
    return mergedcollections
end

function randomlocations(n::Int)::Vector{Location}
    v = Vector{Location}(undef, n)
    r = rand(1:1000, 2 * n, 1)
    for i = 1:n
        longi = r[i]
        lati = r[i+n]
        v[i] = Location(longi, lati)
    end
    return v
end


@testset "Distance" begin
    @test distance(Location(0, 0), Location(3, 4)) == 5
    @test distance(Location(0, 0), [Location(3, 4), Location(100, 100)]) == 5
    @test distance([Location(3, 4), Location(100, 100)], Location(0, 0)) == 5
    @test distance([Location(0, 100), Location(0, 0)], [Location(3, 4), Location(100, 100)]) == 5
end

@testset "Clustering" begin
    @test length(clustering([Location(0, 0), Location(1,1), Location(1, 10), Location(1, 11)], 2)) == 2
end

@test pathlength([Location(0, 0), Location(3, 4), Location(0, 0)]) == 10



@test length(randomlocations(5)) == 5